File: frameworks/persistence.spring/src/main/java/com/example/scaffold/frameworks/persistence/spring/SpringPagination.java
package com.example.scaffold.frameworks.persistence.spring;

import com.example.frameworks.domain.core.Pagination;
import com.example.frameworks.domain.core.Sort;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

public class SpringPagination implements Pagination {

    private final Pageable impl;

    public static Pagination of(PageRequest impl) {
        return new SpringPagination(impl);
    }

    protected SpringPagination(Pageable pageable) {
        this.impl = pageable;
    }

    @Override
    public int pageNumber() {
        return impl.getPageNumber();
    }

    @Override
    public int pageSize() {
        return impl.getPageSize();
    }

    @Override
    public @NotNull Sort sort() {
        return SpringSort.of(impl.getSort());
    }

}

File: frameworks/persistence.spring/src/main/kotlin/com/qomolangma/frameworks/persistence/spring/SpringPagination.kt
package com.qomolangma.frameworks.persistence.spring

import com.qomolangma.frameworks.domain.core.Pagination
import com.qomolangma.frameworks.domain.core.Sort
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable

open class SpringPagination protected constructor(private val impl: Pageable) : Pagination {
    override fun pageNumber(): Int {
        return impl.pageNumber
    }

    override fun pageSize(): Int {
        return impl.pageSize
    }

    override fun sort(): Sort {
        return SpringSort.of(impl.sort)
    }

    companion object {
        @JvmStatic
        fun of(impl: PageRequest): Pagination {
            return SpringPagination(impl)
        }
    }
}
